### ISSUE TITLE
[MJAVADOC-618] Goal javadoc:aggregate fails with submodules packaged as war

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-javadoc-plugin#910]
**[Andrea Del Bene](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bitstorm)** opened **[MJAVADOC-618](https://issues.apache.org/jira/browse/MJAVADOC-618?redirect=false)** and commented

If we have a submodule project producing a war archive, `javadoc:aggregate` fails with the following message:

```
[WARNING] Only outputDirectories and jars are accepted on the path
 [ERROR] Error while creating javadoc report: Only outputDirectories and jars are accepted on the path
 java.lang.IllegalArgumentException: Only outputDirectories and jars are accepted on the path
     at org.codehaus.plexus.languages.java.jpms.LocationManager.resolvePath (LocationManager.java:333)
     at org.codehaus.plexus.languages.java.jpms.LocationManager.resolvePath (LocationManager.java:139)
     at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.addJavadocOptions (AbstractJavadocMojo.java:5001)
```

After some investigations it seems it tries to pass the war package to LocationManager.resolvePath.

I've added a quickstart project to reproduce the problem. Just run:

`mvn package javadoc:aggregate`

---

**Affects:** 3.1.0, 3.1.1

**Attachments:**
- [maven-modular.zip](https://issues.apache.org/jira/secure/attachment/12975527/maven-modular.zip) (_37.47 kB_)

**Issue Links:**
- [MJAVADOC-632](https://issues.apache.org/jira/browse/MJAVADOC-632) agggregate using jpms and non jar module is failing
 (_**"is duplicated by"**_)

0 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), I expect this not to fail, should it?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

So this one is actually a tricky usecase. As mentioned in the error, the module path only allows jars and outputDirectories, not wars or zips.
And there's another issue: there's no specifications for webarchives regarding module descriptors or module names, it has only been specified for jars.
[simonebordet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=simonebordet) are you aware of any issue or JEP to solve this for webarchives?

Based on that it will be quite a challenge to solve this. It might work if we manage to get the output directory in case of non-jar files. So it'll only work if they are part of the Maven multimodule.

**Comment by jira-importer** [comment]
**[Simone Bordet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sbordet)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) I am not aware of any JEP related to JPMS module descriptors for war files. This would have to be defined at Eclipse in JakartaEE.

Perhaps you can trick the javadoc plugin into thinking that it is a jar file with a made-up `Automatic-Module-Name`, or feed the `src/main/java` directory as if it were a jar file?

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

I just  tried different options and I do not see how to include javadoc from war maven module in the jpms world.

Using --patch-module mymodule=pathtomywar doesn't work as javadoc expect module-info.class to be in root directory and not in WEB-INF/classes

ATM users can simply exclude this module from aggregated javadoc.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [0d0e0cc7325fc27b89b7bde0ba2cbe322ff6a040](https://gitbox.apache.org/repos/asf?p=maven-javadoc-plugin.git;a=commit;h=0d0e0cc7325fc27b89b7bde0ba2cbe322ff6a040)
In case of a modular aggregation report, if the artifact is not a file ending with `.jar` the output directory will be used.
This implies, that such module cannot have an automatic module name (these are added to the jar during packaging), so it must contain a module-info.java file.
